### PR TITLE
feat: new api with custom scope and key, filtering, more UI customiza…

### PR DESCRIPTION
**Pull request by npaton**
This contains a lot of changes, that were utterly untested.

I've been struggling to get my local setup to build, and rather than spend another hour on that, I decided to make the changes to show what the new API would look like.

So we have new props:
- scope: which is where to attach the chat: game, round or stage
- key: which is the key to use to save the chat on the scope (defaults to "chat")
- filter: to filter messages to be shown
- onNewMessage: to modify or dismiss a message before it's created
- and some UI components overrides:
  - header: the open chat window header
  - closed: the closed chat window header
  - message: the message
  - footer: the footer of the chat window that contains the input

I did not add back the floating version of the chat window. That might be something we still want.

@hubertusputu1 Let me know if you are having problems with any of the changes.
@JamesPHoughton Please check out the API, see if it answers all your needs (except maybe the floating window, @hubertusputu1 maybe that's something you can add back from James' PR).

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -27,12 +27,12 @@ import { Chat } from "@empirica/chat";
 
 ### Multiple chat instances within the same scope
 
-You can pass an optional `key` string prop to differenciate different chats
-within the same scope. This changes which key on the given scope the chat will
+You can pass an optional `customKey` string prop to differenciate different chats
```
Comment (npaton) [inline Readme.md:33]: Oh yeah, lol, oops, `key` is reserved in React. `customKey` sounds good 👍

File: src/Messages.js [hunk 2]
```diff
@@ -12,31 +12,28 @@ export default class Messages extends React.Component {
...
-        ) : null}
+        {messages.length === 0 ? <div className="empty">No messages yet...</div> : null}
         {messages.map((message, i) => {
           return (
-            <MessageComp
-              key={i}
-              message={message}
-              self={message.playerId === player._id}
-            />
+            <MessageComp key={i} message={message} player={player} self={message.playerId === player._id} />
```
Comment (npaton) [inline src/Messages.js]: `self={...}` is no longer needed here, it's done inside the component now.

File: Readme.md [hunk 3]
```diff
@@ -27,12 +27,12 @@ import { Chat } from "@empirica/chat";
...
+within the same scope. This changes which customKey on the given scope the chat will
```
Comment (npaton) [inline Readme.md]: maybe call it "get/set key" in this context.

Comment (hubertusputu1) [reply Readme.md]: something like this "You can get/set key as `customKey` string ...... " ?

Comment (npaton) [reply Readme.md]:
```suggestion
within the same scope. This changes which get/set key on the given scope the chat will
```

**Comment by hubertusputu1** [comment]
note: i am still adding the floating window on here, and then will do another test.

**Comment by npaton** [comment]
@JamesPHoughton Did you get a chance to review the new design: https://github.com/empiricaly/chat/blob/3a67ded4901af0580387dadb359ab48f312eedd4/Readme.md

Does it answer all your needs?

**Comment by JamesPHoughton** [comment]
I haven't had a chance to give it a go yet, but just from the API I think it should. 

As it is, is the best way to float the window to overwrite the chat package CSS from the empirical project? I need to learn more about how to do that...

One question - with the current CSS, if there are multiple chat windows, will they stack on top of one another, or do they tile?

**Comment by npaton** [comment]
Good question. @hubertusputu1 Let's make sure they don't stack when fixed.
And for "floating", I wasn't sure if this means like you can drag the window around, or you can embed the chat somewhere in the code of the page (which is currently what's supported I believe).

**Comment by JamesPHoughton** [comment]
Good point, "floating" is ambiguous. All I need to do is to embed the chat window in the page, not actually to let the user move it around.

**Comment by npaton** [comment]
Ok, thanks. We'll make sure that's easy to do then. 👍

**Comment by JamesPHoughton** [comment]
Trying this out, it looks like the default behavior doesn't display the name field of other chat participants at all:
```jsx
<Chat
  scope={round}
  player={player}
  customKey={chatId}
/>
```

image


Is this expected?

**Comment by JamesPHoughton** [comment]
It's because I have no `player.name` attribute defined. Are we suggesting that this be a standard pattern for empirica games, to use that in place of `player.id` when a public-facing identifier (i.e. either a pseudonym or real name) is called for?

**Comment by JamesPHoughton** [comment]
I think i prefer @amaatouq's suggestion: https://github.com/empiricaly/chat/pull/1#issuecomment-605073257
Updated here: https://github.com/JamesPHoughton/chat/tree/displayName


(Never done a PR to a PR before? Lets try it!)

**Comment by JamesPHoughton** [comment]
Also, the chat window is launching "closed" because the `isOpen` state is not initialized, but `isChatOpen` is used in some places and `isOpen` is used in others. Switched all to `isChatOpen` but no real preference for this over `isOpen`. fixed in my branch.

**Comment by JamesPHoughton** [comment]
nope, I misunderstood how things were working! You're adding `name` attribute to the player dictionary stored with the message, which I like better than fetching the name for the player at display time. I just needed to set the name object on the players. Sorry for all this confusion...

**Comment by npaton** [comment]
Thank you for the update!

For the name, I was wondering if we should have a default, like show their Player._id (or the first 6 characters of it), so at least there's something. And maybe at the same time, in dev, log to the console that "name" should be set on the player.

**Comment by JamesPHoughton** [comment]
> For the name, I was wondering if we should have a default, like show their Player._id (or the first 6 characters of it) ... And maybe at the same time, in dev, log to the console that "name" should be set on the player.

That works for me, or just fill in "player 1", etc. and throw the warning. I don't think we want experimenters to depend on the default.

**Comment by JamesPHoughton** [comment]
Also, having a bug where the chat window doesn't scroll properly (ie, scroll to bottom after new messages are added). The reason seems to be because the function that would set `scrollTop = scrollHeight` is depending on the old syntax in which messages are passed into the `Messages` object:

```jsx
componentDidUpdate(prevProps) {
    const { messages: prevMessages } = prevProps;
    const { messages: currentMessages } = this.props

    if (prevMessages && currentMessages && (prevMessages.length < currentMessages.length)) {
      this.messagesEl.current.scrollTop = this.messagesEl.current.scrollHeight;
    }
  }
```

Problem is, there is no `messages` object in props anymore, as were `scope.get()`-ing them. So, the `if` statement is never `true`.

It seems like a straightforward workaround is to store the length of the message array in the `Messages` state, and check it against what we `get`. 

I'll put this together and see how it goes, if you have a better pattern for it @npaton, we can do that instead.

It seems like the `componentDidUpdate` method is getting called a lot - if it gets called as frequently as `render()` maybe we just put the scroll control in render just to save the double `get`?

**Comment by JamesPHoughton** [comment]
[updates in my fork](https://github.com/JamesPHoughton/chat/commit/9a66834a06f90a4ac1c9233135353fed25fd7605) fix the scrolling issue. I put changes in `render()` to avoid replicating effort on getting the messages. @npaton @amaatouq, If you like this I'll remove the commented out `componentDidUpdate` and PR.

**Comment by npaton** [comment]
Hey @JamesPHoughton Sorry for the slow reply. Thank you for testing this further and finding this issue!!

It is frowned upon updating state from render(), as it could lead to an infinite loop, so I would avoid that. It seems unlikely here, but it's still a bad idea overall. That said, `num_messages` is not reactive from what I can see, so there is no reason to have it in state. Even if the messages array length comparison does not use this.state, doing something on elements directly like scrolling, should not happen inside of render(). I am not sure why componentDidUpdate is called so much, but it shouldn't be more than the input props are changing (use PureComponent?), we should look into that. And keep the scroll trigger in componentDidUpdate, even if we have to filter twice. Ideally we'd calculate it once and store it, though with the later react APIs this has become more complicated without componentWillUpdate(), so you'd preferably need to do it above Messages, could have data component that does just that. Maybe 

```jsx
function filteredMessages(WrappedComponent) {
  return class extends React.PureComponent {
    render() {
      const { scope, customKey, filter, ...rest } = this.props;

      let messages = scope.get(customKey) || [];
      if (filter) {
        messages = filter(messages);
      }
  
        return <WrappedComponent messages={messages} {...rest} />;
    }
  };
}

class Messages extends React.PureComponent {
    // ...
}

export default filteredMessages(Messages)
```

I do not have time to look at this just now. @hubertusputu1 Would you mind looking into this? Once done, let's do a final review and close this soon.

**Comment by JamesPHoughton** [comment]
Thanks @npaton, good to know about render not updating state. Not sure if I'll have time in the next couple days to give it another go, so @hubertusputu1 , feel free to make changes.

Also, It seems that what I have isn't fixing the issue completely. After the scroll gets updated, the last message still isn't visible - it seems to be sitting under the text entry box. You can scroll down to see it, but the autoscroll I have doesn't scroll all the way down. Not sure if it's a mistake I'm making or if its something to fix with a css pad or adding a little bit to the scroll height.

**Comment by hubertusputu1** [comment]
@npaton @JamesPHoughton sure, i will take a look on it.

**Comment by npaton** [comment]
LGTM. Awesome, thank you @hubertusputu1. If scrolling is confirmed to work, please merge and release. 🙌 

> Also, It seems that what I have isn't fixing the issue completely. After the scroll gets updated, the last message still isn't visible - it seems to be sitting under the text entry box. You can scroll down to see it, but the autoscroll I have doesn't scroll all the way down. Not sure if it's a mistake I'm making or if its something to fix with a css pad or adding a little bit to the scroll height.

@JamesPHoughton this was likely due to the scroll happening in render(). At the time of render being called, the new message hasn't been inserted into the DOM yet, so the offset is incorrect. In `componentDidUpdate()` we are after the result of render has been applied to the DOM, the new message is in there, so the scroll offset is then correct. 👍

**Comment by JamesPHoughton** [comment]
>this was likely due to the scroll happening in render()

Thanks @npaton, that makes sense! 

@hubertusputu1, the autoscroll is still not working for me, I'm afraid. It seems that in the `componentDidUpdate`, `prevMessages` does not differ from `currentMessages`, even when new messages are added.

Testing with:
```jsx
componentDidUpdate(prevProps) {
    const { messages: prevMessages } = prevProps;
    const { messages: currentMessages } = this.props;
    console.log(prevMessages.length == currentMessages.length)  // add test

    if (this.messagesEl.current !== null && currentMessages.length > prevMessages.length) {
      this.messagesEl.current.scrollTop = this.messagesEl.current.scrollHeight;
    }
  }
```
always prints `true`. I'm not sure quite why, see what I can find.

**Comment by JamesPHoughton** [comment]
A guess is that the messages array is just getting mutated, and so it's comparing the array to itself. If I make a copy before passing it as a prop to `Messages` the scrolling works as expected:

```jsx
function filteredMessages(WrappedComponent) {
  return class extends React.PureComponent {
    render() {
      const { scope, customKey, filter, ...rest } = this.props;
      let messages = scope.get(customKey) || [];
      if (filter) {
        messages = filter(messages);
      }
      const copy_messages = [...messages];  // make a copy
      return <WrappedComponent messages={copy_messages} {...rest} />;
    }
  };
}
```

**Comment by hubertusputu1** [comment]
@JamesPHoughton can you please try to installing a new package from `npm i @empirica/chat` on your project? Let's see if the auto-scrolling works or no.

**Comment by npaton** [comment]
@hubertusputu1 So this is working for you? @JamesPHoughton But it's not working for you?
If so, @JamesPHoughton, can you tell us about the context you're testing (browsers, steps to reproduce if any). Thanks!

**Comment by JamesPHoughton** [comment]
Yes! I've added you both as collaborators on our experiment repo, so you can have a look (the game directory is under 'telephone_game'. This is running using the meteor local server, and I'm using chrome `Version 81.0.4044.122 (Official Build) (64-bit)`. I'm using  https://github.com/empiricaly/meteor-empirica-core/commit/92903252bdda34ddd62aade459a580517dc87565

If you want to reproduce, change `client/chat/src/Messages.js` line 13 from:
``` jsx
return <WrappedComponent messages={[...messages]} {...rest} />;
```
to
```jsx
return <WrappedComponent messages={messages} {...rest} />;
```

**Pull request merged**